shell-completion/Makefile

```diff
@@ -1,30 +1,34 @@
-
-.PHONY: command-options clean
+# Setting the number of job runners like this in the makefile only works in
+# GNU Make 4.3 or later. Older versions will require that either an env
+# variable be set before running or command line flag be passed at runtime to
+# get parallel jobs.
+MAKEFLAGS += --jobs=$(shell nproc 2>/dev/null || printf 8)
 
 all: hledger-completion.bash
 
-hledger-completion.bash: hledger-completion.bash.m4 commands-list.txt query-filters.txt generic-options.txt command-options
+COMMANDS := $(sort $(shell hledger | ./output-commands.sh | grep -v ^hledger | sort -u) ui web api)
+CMDOPTFILES := $(foreach CMD,$(COMMANDS),options-$(CMD).txt)
+
+hledger-completion.bash: hledger-completion.bash.m4 commands.txt commands-list.txt query-filters.txt generic-options.txt $(CMDOPTFILES)
 	m4 hledger-completion.bash.m4 > $@
 
 generic-options.txt:
 	hledger -h | ./output-options.sh | sort -u > $@
 
 commands.txt:
-	hledger | ./output-commands.sh | grep -v ^hledger | sort -u > $@
-	echo ui  >> $@
-	echo web >> $@
-	echo api >> $@
+	printf "%s\n" $(COMMANDS) > $@
 
-commands-list.txt: commands.txt
-	paste -sd, $^ | tr -d '\n' > $@
+commands-list.txt:
+	sed -e 's/ /,/g' <<< "$(COMMANDS)" > $@
```

Comment:
I forgot heredocs always end in a newline (long standing pet peeve of mine). Fixed with sed to avoid a new process. Any of the solutions work, but I hate `paste` (confusing syntax and misleading name) and the `make` convolutions are confusing to the uninitiated. As long as we're running one shell process (as `printf` would be) and we already require `sed` elsewhere in the project I figure we can keep it to just that one process. At the end of the day that's just preference though, as you note any of these 4 ways would work.